### PR TITLE
Add mvx bottom navigation view

### PR DESCRIPTION
**Pull request by paulppn**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Feature

### :arrow_heading_down: What is the current behavior?
Showing the bottom navigation can't be navigated using page presentation hints

### :new: What is the new behavior (if this is a feature change)?
Introduced the MvxBottomNavigationView (on Android), so we can create a bottom navigation which will listen to page presentation hints. It can be populated the same way a tab layout can be populated,  using attributes and a ShowInitialViewModelsCommand. To do this I added a MvxBottomNavigationViewPresentationAttribute.

### :boom: Does this PR introduce a breaking change?
No

### :bug: Recommendations for testing
See the added sample in the Playground

### :memo: Links to relevant issues/docs
https://github.com/MvvmCross/MvvmCross/issues/2358
https://github.com/MvvmCross/MvvmCross/issues/2114

### :thinking: Checklist before submitting

- [x] All projects build
- [x] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [ ] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contributing/mvvmcross-docs-style-guide))
- [x] Rebased onto current develop

### CODE REVIEW

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+using Android.Util;
+using Android.Views;
+using Android.Support.Design.Widget;
+
+namespace MvvmCross.Droid.Support.Design
+{
+    [Register("mvvmcross.droid.support.design.MvxBottomNavigationView")]
+    public class MvxBottomNavigationView : BottomNavigationView, BottomNavigationView.IOnNavigationItemSelectedListener
+    {
+        public List<MvxMenuItem> ItemsSource = new List<MvxMenuItem>();
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Use ItemsSource like this: https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross/Platforms/Android/Binding/Views/MvxListView.cs#L73

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            return ItemsSource.FindIndex(i => i.ViewModel == viewModel);
+        }
+
+        public MvxMenuItem FindItemByMenuItem(IMenuItem item)
+        {
+            return ItemsSource.Find(i => i.MenuItem == item);
+        }
+
+        public ICommand HandleNavigate { get; set; }
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Should be more like: https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross/Platforms/Android/Binding/Views/MvxListView.cs#L85

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public class MvxMenuItem
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Move into it's own file

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        public IMenuItem MenuItem { get; set; }
+        public Type ViewModel { get; set; }
+
+        public MvxMenuItem(IMenuItem menuItem, Type viewModel)
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Maybe this needs to be an interface instead, then people can have their own custom class with lots of info. Then we can pass that to the Command?

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+namespace MvvmCross.Droid.Support.Design
+{
+    [Register("mvvmcross.droid.support.design.MvxBottomNavigationView")]
+    public class MvxBottomNavigationView : BottomNavigationView, BottomNavigationView.IOnNavigationItemSelectedListener
+    {
+        public List<MvxMenuItem> ItemsSource = new List<MvxMenuItem>();
+
+        public MvxBottomNavigationView(Context context) : base(context)
+        {
+            this.SetOnNavigationItemSelectedListener(this);
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Maybe implement dispose and remove the listener?

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        public void AddItem(IMenuItem item, Type viewModel)
+        {
+            ItemsSource.Add(new MvxMenuItem(item, viewModel));
+
+            // The first item is autoselected
+            if (ItemsSource.Count == 1)
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Should this be equal or bigger then?

File: MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                OnNavigationItemSelected(item);
```
Comment (martijn00) [inline MvvmCross.Android.Support/Design/MvxBottomNavigationView.cs]: Maybe we need to make a bool that defaults to true if the user wants to auto open the first tab?

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 8]
```diff
@@ -104,6 +106,14 @@ public override MvxBasePresentationAttribute GetPresentationAttribute(MvxViewMod
                                 break;
                             }
 
+                            // check for MvxBottomNavigationViewPresentationAttribute 
+                            if (item is MvxBottomNavigationViewPresentationAttribute mvxBottomNavigationAttribute
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Don't use mvx prefix in local variables

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 9]
```diff
@@ -167,6 +177,38 @@ public override Task<bool> ChangePresentation(MvxPresentationHint hint)
                         return Task.FromResult(true);
                     }
                 }
+                else
+                {
+                    if (attribute is MvxBottomNavigationViewPresentationAttribute mvxBottomNavigationViewAttribute)
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Don't use mvx prefix in local variables

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 10]
```diff
@@ -167,6 +177,38 @@ public override Task<bool> ChangePresentation(MvxPresentationHint hint)
...
+                    {
+                        var mvxBottomNavigationView = FindMvxBottomNavigationViewInFragmentPresentation(mvxBottomNavigationViewAttribute);
+                        if (mvxBottomNavigationView != null)
+                        {
+                            var mvxMenuItem = mvxBottomNavigationView.FindItemByViewModel(pagePresentationHint.ViewModel);
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Don't use mvx prefix in local variables

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 11]
```diff
@@ -167,6 +177,38 @@ public override Task<bool> ChangePresentation(MvxPresentationHint hint)
...
+
+                            if (mvxMenuItem == null)
+                            {
+                                //MvxAndroidLog.Instance.Trace("Did not find a menu item which corresponds to {0}, skipping presentation change...",
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Add the log back in

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 12]
```diff
@@ -192,6 +234,26 @@ private ViewPager FindViewPagerInFragmentPresentation(MvxViewPagerFragmentPresen
             return viewPager;
         }
 
+        private MvxBottomNavigationView FindMvxBottomNavigationViewInFragmentPresentation(MvxBottomNavigationViewPresentationAttribute pagerFragmentAttribute)
+        {
+            MvxBottomNavigationView MvxBottomNavigationView = null;
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Naming convention

File: MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs [hunk 13]
```diff
@@ -503,6 +565,26 @@ protected override void ShowNestedFragment(Type view, MvxFragmentPresentationAtt
             return true;
 
         }
+
+        protected virtual Task<bool> ShowBottomNavigationFragment(
+            Type view,
+            MvxBottomNavigationViewPresentationAttribute attribute,
+            MvxViewModelRequest request)
+        {
+            var mvxBottomNavigationView = FindMvxBottomNavigationViewInFragmentPresentation(attribute);
```
Comment (martijn00) [inline MvvmCross.Android.Support/V7.AppCompat/MvxAppCompatViewPresenter.cs]: Don't use mvx prefix in local variables

**Comment by martijn00** [comment]
In general i think this could also be used in the `NavigationView` since that uses IMenuItem as well. The implementation for that would be a `MvxNavigationView` which would look very similar to the MvxBottomNavigationView.

**Comment by nickrandolph** [comment]
@paulppn can you force push after you rebase please to eliminate the excess commits

**Comment by mierzynskim** [comment]
Are there any plans to release it soon? It would be useful in my project 😄. I can help if there's anything missing in this PR.

**Comment by Hackmodford** [comment]
@Cheesebaron Is there a way I can get permission to modify this PR?

**Comment by paulppn** [comment]
> @Cheesebaron Is there a way I can get permission to modify this PR?

I added you as a collaborator on my fork, so I guess you should be able to edit the branch, if there is anything else you need me to do, please let me know.

**Comment by Hackmodford** [comment]
I'm rebasing off the latest dev branch. This had the side effect of some files being removed and then me recreating them in the new location. Hope that's okay.

**Comment by Hackmodford** [comment]
when I try to push to your branch I get this.

`Permission to paulppn/MvvmCross.git denied to Hackmodford.`

**Comment by Hackmodford** [comment]
Apparently I'm a moron. I forgot you need to accept the invite 😆

Could someone confirm the playground works? I'm unable to get the project to build on macOS.

**Comment by Hackmodford** [comment]
I think I just came up with simpler implementation that uses a ViewPager.

As far as the user is concerned, it works almost the same as the MvxTabLayoutPresentation. The only difference is you have to use a MvxBottomNavigationView.

@Cheesebaron Would it be better to create a new Merge Request or update this one?

**Comment by Hackmodford** [comment]
Decided to push to this branch.

This implementation basically works the same. 

Added a `MvxBottomNavigationViewPresentationAttribute`
Added a `MvxBottomNavigationView`
This view has a reference to the ViewPager and handles swiping between pages and updating the bottomNavigationView.

MenuItems ids are set to their index in the viewPager, this allows the bottomNavigationView to understand which page should be selected instead of relying on the ViewModel to handle this.

This implementation allows swiping between pages because of the view pager :)

Uses the existing AndroidViewPresenter logic for displaying fragments in a view pager. This should make it easier to maintain.

**Pull request closed without merging**